### PR TITLE
[Merged by Bors] - feat (NumberTheory/LSeries): Hurwitz zeta

### PR DESCRIPTION
**Pull request by loefflerd**
Add the Hurwitz zeta function (defined as the sum of its even and odd parts, which were already defined in previous PR's).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+  `∑' n : ℕ, exp (2 * π * I * n * a) / n ^ s`.
+
+## Main definitions and results
+
+* `hurwitzZeta`: the Hurwitz zeta function (defined to be periodic in `a` with period 1)
+* `expZeta`: the exponential zeta function
+* `hasSum_nat_hurwitzZeta_of_mem_Icc` and `hasSum_expZeta_of_one_lt_re`:
+  relation to Dirichlet series for `1 < re s`
+* `differentiableAt_hurwitzZeta` and `differentiableAt_expZeta`: analyticity away from `s = 1`
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:33]:
```suggestion
* ` hurwitzZeta_residue_one` shows that the residue at `s = 1` equals `1`
* `differentiableAt_hurwitzZeta` and `differentiableAt_expZeta`: analyticity away from `s = 1`
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+/-- The Hurwitz zeta function, which is the meromorphic continuation of
+`∑ (n : ℕ), 1 / (n + a) ^ s` if `0 ≤ a ≤ 1`. See `hasSum_hurwitzZeta_of_one_lt_re` for the relation
+to the Dirichlet series in the convergence range. -/
+noncomputable def hurwitzZeta (a : UnitAddCircle) (s : ℂ) :=
+  hurwitzZetaEven a s + hurwitzZetaOdd a s
+
+lemma hurwitzZetaEven_eq (a : UnitAddCircle) (s : ℂ) :
+    hurwitzZetaEven a s = (hurwitzZeta a s + hurwitzZeta (-a) s) / 2 := by
+  simp_rw [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:51]:
```suggestion
  simp only [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+  ring_nf
+
+lemma hurwitzZetaOdd_eq (a : UnitAddCircle) (s : ℂ) :
+    hurwitzZetaOdd a s = (hurwitzZeta a s - hurwitzZeta (-a) s) / 2 := by
+  simp_rw [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:56]:
```suggestion
  simp only [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    HasSum (fun n : ℕ ↦ 1 / (n + a : ℂ) ^ s) (hurwitzZeta a s) := by
+  convert (hasSum_nat_hurwitzZetaEven_of_mem_Icc ha hs).add
+      (hasSum_nat_hurwitzZetaOdd_of_mem_Icc ha hs) using 1
+  ext1 n
+  ring_nf
+
+/-- The residue of the Hurwitz zeta function at `s = 1` is `1`. -/
+lemma hurwitzZeta_residue_one (a : UnitAddCircle) :
+    Tendsto (fun s ↦ (s - 1) * hurwitzZeta a s) (𝓝[≠] 1) (𝓝 1) := by
+  simp_rw [hurwitzZeta, mul_add, (by simp : 𝓝 (1 : ℂ) = 𝓝 (1 + (1 - 1) * hurwitzZetaOdd a 1))]
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:76]:
```suggestion
  simp only [hurwitzZeta, mul_add, (by simp : 𝓝 (1 : ℂ) = 𝓝 (1 + (1 - 1) * hurwitzZetaOdd a 1))]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    DifferentiableAt ℂ (fun s ↦ hurwitzZeta a s - 1 / (s - 1) / Gammaℝ s) 1 := by
+  simp_rw [hurwitzZeta, add_sub_right_comm]
+  exact (differentiableAt_hurwitzZetaEven_sub_one_div a).add (differentiable_hurwitzZetaOdd a 1)
+
+/-- Expression for `hurwitzZeta a 1` as a limit. (Mathematically `hurwitzZeta a 1` is
+undefined, but our construction assigns some value to it; this lemma is mostly of interest for
+determining what that value is). -/
+lemma tendsto_hurwitzZeta_sub_one_div_nhds_one (a : UnitAddCircle) :
+    Tendsto (fun s ↦ hurwitzZeta a s - 1 / (s - 1) / Gammaℝ s) (𝓝 1) (𝓝 (hurwitzZeta a 1)) := by
+  simp_rw [hurwitzZeta, add_sub_right_comm]
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:90]:
```suggestion
  simp only [hurwitzZeta, add_sub_right_comm]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  refine (tendsto_hurwitzZetaEven_sub_one_div_nhds_one a).add ?_
+  exact (differentiable_hurwitzZetaOdd a 1).continuousAt.tendsto
+
+/-- The difference of two Hurwitz zeta functions is differentiable everywhere. -/
+lemma differentiable_hurwitzZeta_sub_hurwitzZeta (a b : UnitAddCircle) :
+    Differentiable ℂ (fun s ↦ hurwitzZeta a s - hurwitzZeta b s) := by
+  simp_rw [hurwitzZeta, add_sub_add_comm]
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:97]:
```suggestion
  simp only [hurwitzZeta, add_sub_add_comm]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+/-- If `a ≠ 0` then the exponential zeta function is analytic everywhere. -/
+lemma differentiable_expZeta_of_ne_zero {a : UnitAddCircle} (ha : a ≠ 0) :
+    Differentiable ℂ (expZeta a) :=
+  (differentiableAt_expZeta a · (Or.inr ha))
+
+/-!
+## The functional equation
+-/
+
+/-- Functional equation for Hurwitz zeta function. -/
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:143]:
```suggestion
/-- Functional equation for the Hurwitz zeta function. -/
```

File: Mathlib/NumberTheory/LSeries/HurwitzZeta.lean [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+lemma hurwitzZeta_one_sub (a : UnitAddCircle) {s : ℂ}
+    (hs : ∀ (n : ℕ), s ≠ -n) (hs' : a ≠ 0 ∨ s ≠ 1) :
+    hurwitzZeta a (1 - s) = (2 * π) ^ (-s) * Gamma s *
+    (exp (-π * I * s / 2) * expZeta a s + exp (π * I * s / 2) * expZeta (-a) s) := by
+  rw [hurwitzZeta, hurwitzZetaEven_one_sub a hs hs', hurwitzZetaOdd_one_sub a hs,
+    expZeta, expZeta, Complex.cos, Complex.sin, sinZeta_neg, cosZeta_neg]
+  ring_nf
+
+/-- Functional equation for exponential zeta function. -/
```
Comment (faenuccio) [inline Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:152]:
```suggestion
/-- Functional equation for the exponential zeta function. -/
```

**Comment by faenuccio** [comment]
LGTM!

**Comment by loefflerd** [comment]
Thanks for the review, Filippo! I have applied your changes. I also fixed a minor formatting glitch in the module docstring, and added a mssing lemma `LSeriesHasSum_exp` that makes the link with Michael's `LSeries` API.

**Comment by faenuccio** [comment]
> Thanks for the review, Filippo! I have applied your changes. I also fixed a minor formatting glitch in the module docstring, and added a mssing lemma `LSeriesHasSum_exp` that makes the link with Michael's `LSeries` API.

Great!

**Comment by loefflerd** [comment]
Hi Michael,

Thanks for the suggestions! I'm in two minds about this sort of thing (trading off concision / elegance vs. speed). CPU cycles are cheap, after all; and there is the very good point raised in the Zulip discussion "To automate or not not automate?", that making explicit workarounds for slow tactics makes it less likely that tactic-development people will notice and work on the slowness. 

Hence I've put in your `generalize` changes, but with a comment pointing out that it would be nice if `ring_nf` could do these things automatically. Are you happy with that compromise?

Regards, David

**Comment by MichaelStollBayreuth** [comment]
I do understand that (and my suggestions were meant as such, not as requirements).
I think the compromise (adding comments) is reasonable.
Of course, a better solution would be to make `ring_nf` faster (along similar lines, if at all  possible), but this is beyond my paygrade...

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM!

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**